### PR TITLE
Add dropout support for SMPC

### LINKED ISSUE #2500: Dropout is currently not supported for SMPC

### PR DESCRIPTION
**Pull request by sukhadj**
fixes #2500 .
Currently using dropout function specified on ref. https://stackoverflow.com/questions/54109617/implementing-dropout-from-scratch. 
More robust implementation to be added.
This code block is yet to be tested.

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 1]
```diff
@@ -845,6 +845,31 @@ def linear(*args):
...
+                    if training:
+                        binomial = torch.distributions.binomial.Binomial(probs=1 - p)
+
+                        # we must convert the normal tensor to fixed precision before multiplication
+                        noise = (
+                            (
+                                binomial.sample(input.shape).type(torch.FloatTensor)
+                                * (1.0 / (1.0 - p))
+                            )
+                            .fix_prec()
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: Here put `.fix_prec(**input.get_class_attributes())`  to get the same field & precission  than te input otherwise you won't be able to multiply them in the more general cases

File: syft/frameworks/torch/tensors/interpreters/precision.py [hunk 2]
```diff
@@ -845,6 +845,31 @@ def linear(*args):
...
+                    Ref:  https://stackoverflow.com/questions/54109617/implementing-dropout-from-scratch
+                    """
+                    if training:
+                        binomial = torch.distributions.binomial.Binomial(probs=1 - p)
+
+                        # we must convert the normal tensor to fixed precision before multiplication
+                        noise = (
+                            (
+                                binomial.sample(input.shape).type(torch.FloatTensor)
+                                * (1.0 / (1.0 - p))
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/precision.py]: Quick question here, shouldn't you multiply by `p` instead ?
I'm no dropout expert but I just read this: https://machinelearningmastery.com/dropout-for-regularizing-deep-neural-networks/

**Comment by sukhadj** [comment]
> Don't forget to add a test:
> 
>     * at training you might just want to show the signal decrease
> 
>     * at testing just show the signal is unchanged

Done.
This test just checks if dropout has filtered out some inputs.
A more robust check would be comparing `(num_of_inputs_filtered / input_size)` to p. The first vale is close to p. So should we put a threshold on `((num_of_inputs_filtered / input_size) - p)` ? Thoughts on this?

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 1 above)
Comment (sukhadj) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Done :+1:

File: syft/frameworks/torch/tensors/interpreters/precision.py (hunk 2 above)
Comment (sukhadj) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: We are using inverted dropout which uses (1/(1-p)) to scale up the result.
Ref: https://www.coursera.org/lecture/deep-neural-network/dropout-regularization-eM33A

But would like to get it reviewed by someone :smiley:

Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/precision.py]: Excellent you're right!
https://pgaleone.eu/deep-learning/regularization/2017/01/10/anaysis-of-dropout/ confirms this

**Comment by sachin-101** [comment]
@sukhadj @LaRiffle #2500 mentions implementing nn.Dropout class for PySyft.
https://github.com/OpenMined/PySyft/blob/0561a5d21c1787afeeb9753a0487f0c91d38a07d/syft/frameworks/torch/nn/rnn.py#L184
But this PR adds it to precision.py
In order to clean fixed_precision #2897 I will export dropout to nn in PR #3253 . @sukhadj  Are there any catchs I should be familiar of ?

**Comment by sukhadj** [comment]
Nothing major!! Go ahead  :+1:
